File: 순열과조합/Permutation.py
n = [1,2,3,4]
r = 3

permutation = []
result = [0] * r
visited = [False] * len(n)

def dfs(n, r, l):
  if l == r:
    permutation.append(result[:])
  else:
    for i in range(len(n)):
      if not visited[i]:
        result[l] = n[i]
        visited[i] = True
        dfs(n, r, l+1)
        visited[i] = False

dfs(n, r, 0)
print(permutation)

File: modules/Permutation.js
const Permutation = (n, r) => {

  if(r === 0) {
    return [];
  }

  if(r < 0) {
    return new Error('second value shoud be greater than 0');
  }

  if(n.length === r) {
    return n;
  }

  if(n.length < r ){
    return new Error("second value shouldn't be greater than first one!");
  }

  const permutation = [];
  const result = new Array(r);
  const visited = (new Array(n.length)).map(ele => false);

  const dfs = (n, r, l) => {
    if(l === r) {
      permutation.push([...result]);
    } else {
      for(let i=0; i<n.length; i++) {
        if (!visited[i]) {
          result[l] = n[i];
          visited[i] = true;
          dfs(n, r, l+1);
          visited[i] = false;
        }
      }
    }
  }

  dfs(n, r, 0);
  return permutation;
}

module.exports = Permutation;